turbopack/crates/turbopack-core/src/module_graph/mod.rs

```diff
@@ -690,8 +702,108 @@ impl SingleModuleGraph {
             }
         }
     }
+
+    /// For each issue computes a (possibly empty) list of traces from the file that produced the
+    /// issue to roots in this module graph.
+    /// There are potentially multiple traces because a given file may get assigned to multiple
+    /// modules depend on how it is used in the application.  Consider a simple utility that is used
+    /// by SSR pages, client side code, and the edge runtime.  This may lead to there being 3
+    /// traces.
+    /// The returned map is guaranteed to have an entry for every issue.
+    pub async fn compute_import_traces_for_issues(
+        &self,
+        issues: &AutoSet<ResolvedVc<Box<dyn Issue>>>,
+    ) -> Result<FxHashMap<ResolvedVc<Box<dyn Issue>>, Vec<ImportTrace>>> {
+        let issue_paths = issues
+            .iter()
+            .map(|issue| async move { Ok((*issue.file_path().await?).clone()) })
+            .try_join()
+            .await?;
+        let mut file_path_to_traces: FxHashMap<FileSystemPath, Vec<ImportTrace>> =
+            FxHashMap::with_capacity_and_hasher(issue_paths.len(), Default::default());
+        // initialize an empty vec for each path we care about
+        for issue in &issue_paths {
+            file_path_to_traces.entry(issue.clone()).or_default();
+        }
+
+        {
+            let modules = self
+                .modules
+                .iter()
+                .map(|(module, &index)| async move {
+                    Ok(((*module.ident().path().await?).clone(), index))
+                })
```

Comment:
How about making `issue_paths` an `FxHashSet` and filtering `modules` by `issue_paths.contain`